htlcswitch/link.go

```diff
@@ -3756,6 +3761,51 @@ func (l *channelLink) sendHTLCError(pd *lnwallet.PaymentDescriptor,
 	)
 }
 
+// sendPeerHTLCFailure handles sending a HTLC failure message back to the
+// peer from which the HTLC was received. This function is primarily used to
+// handle the special requirements of route blinding, specifically:
+// - Forwarding nodes must switch out any errors with MalformedFailHTLC
+// - Introduction nodes should return regular HTLC failure messages.
+//
+// This function expects the correct failure reason to already be set for
+// the introduction node when the failure packet was created.
+//
+// Note: this function does not yet handle special error cases for receiving
+// nodes in blinded paths, as LND does not support blinded receives.
+func (l *channelLink) sendIncomingHTLCFailureMsg(htlcIndex uint64,
+	updateAddBlinding bool,
+	reason lnwire.OpaqueReason) {
+
+	// If a blinding point was populated in update_add_htlc, send a
+	// malformed HTLC error with the route blinding code because we are
+	// not the introduction node.
+	if updateAddBlinding {
+		if err := l.cfg.Peer.SendMessage(
+			false, &lnwire.UpdateFailMalformedHTLC{
+				ChanID:      l.ChanID(),
+				ID:          htlcIndex,
+				FailureCode: lnwire.CodeInvalidBlinding,
+			},
+		); err != nil {
+			l.log.Warnf("Send fail malformed failed: %v", err)
+		}
+
+		return
+	}
+
+	// Otherwise we send a failure message as usual. This case covers both
+	// the introduction point in a blinded path (where we expect our reason
+	// to already have been appropriately set on failure) or just part of a
+	// regular payment path.
+	if err := l.cfg.Peer.SendMessage(false, &lnwire.UpdateFailHTLC{
+		ChanID: l.ChanID(),
+		ID:     htlcIndex,
+		Reason: reason,
```

Comment:
Do we have access to our  `hop.ErrorEncrypter` when we call this from `handleDownstreamPkt`? afaik we do that encryption in the switch (same as malformed / `convertedError`) so by the time we get here we can no longer encrypted the failure reason.

Reply:
there would be `pkt.circuit.ErrorEncrypter`, hope that's the correct one and that it's set? I think it would be great if we could "firewall" all our errors in a single place based on the blinding role. Just a thought, could we extend the `ErrorEncrypter` interface with a `Role` method via `sphinxHopIterator` and `blindingKit`?